### PR TITLE
Replace data

### PR DESCRIPTION
**Pull request by pavel-gfx**
Added functionality for replacing collection data from the sdk

Added methods for creating and uploading to blob storage

[](https://zegami.atlassian.net/browse/ZGM-7215)

### CODE REVIEW

File: zegami_sdk/util.py [hunk 1]
```diff
@@ -108,12 +107,42 @@ def _auth_post(self, url, body, return_response=False, **kwargs):
...
+    """
+    blob_url = f'{self.HOME}/{self.API_1}/project/{workspace_id}/signed_blob_url'
+    blob_id = str(uuid.uuid4())
+    response = self._auth_post(blob_url, body=None, json={"ids": [blob_id]}, return_response=True)
+    try:
+        data = response.json()
+        url = data[blob_id]
+        return url, blob_id
+    except Exception as ex:
+        print(ex)
```
Comment (tarqu1n) [inline zegami_sdk/util.py:134]: Why would we want to to print rather than Raise a custom exception here?

File: zegami_sdk/util.py [hunk 2]
```diff
@@ -108,12 +107,42 @@ def _auth_post(self, url, body, return_response=False, **kwargs):
...
+
+
+def _upload_to_signed_blob_storage(data, url, mime_type, headers=None, **kwargs):
+    """Upload data to an already create blob storage."""
+    if url.startswith("/"):
+        url = f'https://storage.googleapis.com{url}'
+    headers = {'Content-Type': mime_type}
+    if 'windows.net' in url:
+        headers['x-ms-blob-type'] = 'BlockBlob'
```
Comment (tarqu1n) [inline zegami_sdk/util.py:143]: worth a comment here explaining that azure blob storage requires this header and maybe a link to docs

File: zegami_sdk/collection.py [hunk 3]
```diff
@@ -223,6 +235,42 @@ def get_image_urls(self, rows, source=0):
...
+            assert name.split('.')[-1] in ['tsv', 'csv']
+            with open(data, 'r') as f:
+                tsv = f.read()
+
+        zeg_client = self.client
+        upload_dataset_url = f'{zeg_client.HOME}/{zeg_client.API_0}/project/{self.workspace_id}/datasets/{self._upload_dataset_id}'
+        bytes_tsv = bytes(tsv, 'utf-8')
+        mime_type = 'application/octet-stream'
+
+        print(f'Updating collection\'s [id: {self.id}] dataset.\n')
```
Comment (tarqu1n) [inline zegami_sdk/collection.py:259]: Dont think its worth printing here, wasnt sure if this was left in by accident?

File: zegami_sdk/util.py (hunk 1 above)
Comment (ahaith) [reply zegami_sdk/util.py:134]: Agreed. By default, we should just let the exception bubble up to the caller.

File: zegami_sdk/util.py [hunk 4]
```diff
@@ -108,12 +107,36 @@ def _auth_post(self, url, body, return_response=False, **kwargs):
...
     its .json() output.
-
     Any additional kwargs are forwarded onto the requests.put().
     """
     r = requests.put(url, body, headers=self.headers, **kwargs)
     self._check_status(r, is_async_request=False)
     return r if return_response else r.json()
+
+
+def _create_singed_blob_storage(self, workspace_id):
```
Comment (ahaith) [inline zegami_sdk/util.py:119]: Could we rename this to `_obtain_signed_blob_storage_url`?

File: zegami_sdk/util.py [hunk 5]
```diff
@@ -108,12 +107,36 @@ def _auth_post(self, url, body, return_response=False, **kwargs):
...
+    """
+    blob_url = f'{self.HOME}/{self.API_1}/project/{workspace_id}/signed_blob_url'
+    blob_id = str(uuid.uuid4())
+    response = self._auth_post(blob_url, body=None, json={"ids": [blob_id]}, return_response=True)
+    data = response.json()
+    url = data[blob_id]
+    return url, blob_id
+
+
+def _upload_to_signed_blob_storage(data, url, mime_type, headers=None, **kwargs):
```
Comment (ahaith) [inline zegami_sdk/util.py:133]: Not really any such thing as 'signed blob storage'. It's the url that's signed so maybe 'signed blob url', or 'signed blob storage url'.

File: zegami_sdk/collection.py [hunk 6]
```diff
@@ -223,6 +235,40 @@ def get_image_urls(self, rows, source=0):
...
+        The provided input should be a pandas dataframe or a local csv/tsv file.
+        """
+        tsv = ''
+        name = ''
+        if type(data) == pd.DataFrame:
+            tsv = data.to_csv(sep='\t', index=False)
+            name = 'provided_as_dataframe.tsv'
+        else:
+            name = os.path.split(data)[-1]
+            assert name.split('.')[-1] in ['tsv', 'csv']
```
Comment (ahaith) [inline zegami_sdk/collection.py:250]: People will probably want to deal with xlsx files also.

**Review by ahaith — COMMENTED** [review]
Looks good. A few extra small points worth addressing.

File: zegami_sdk/util.py (hunk 4 above)
Comment (pavel-gfx) [reply zegami_sdk/util.py]: > Looks good. A few extra small points worth addressing.

Thanks, Andy, all points have been address in the latest commit.

File: zegami_sdk/collection.py [hunk 7]
```diff
@@ -223,6 +235,48 @@ def get_image_urls(self, rows, source=0):
...
+        else:
+            name = os.path.split(data)[-1]
+            if (file_ext := name.split('.')[-1]) in ['tsv', 'csv']:
+                with open(data, 'r') as f:
+                    tsv = f.read()
+            elif file_ext in ['xls', 'xlsx']:
+                # Not setting the extension to tsv results in error in processing
+                name = name + '.tsv'
+                xlsx_to_pd = pd.read_excel(data, sheet_name=sheet_name)
+                tsv = xlsx_to_pd.to_csv(sep='\t', index=False)
```
Comment (ahaith) [inline zegami_sdk/collection.py:258]: Not necessary to do this conversion here - in fact better not to! The xls/xlsx file can be uploaded as a blob and cloud deals with reading it, the blob then reflects exactly what the user uploaded.

**Review by ahaith — APPROVED** [review]
Looks good. One last thing - please do a minor version bump in setup.py.

**Pull request merged**